src/ints/xms.cpp

```diff
@@ -102,392 +110,752 @@ struct XMS_MemMove{
-		xms_handles[0].free	= false;
-
-		/* Set up UMB chain */
-		umb_available=section->Get_bool("umb");
-		bool ems_available = GetEMSType(section)>0;
-		DOS_BuildUMBChain(section->Get_bool("umb"),ems_available);
-	}
-
-	~XMS(){
-		Section_prop * section = static_cast<Section_prop *>(m_configuration);
-		/* Remove upper memory information */
-		dos_infoblock.SetStartOfUMBChain(0xffff);
-		if (umb_available) {
-			dos_infoblock.SetUMBChainState(0);
-			umb_available=false;
-		}
+	XMS(Section* configuration);
+	~XMS() override;
+};
 
-		if (!section->Get_bool("xms")) return;
-		/* Undo biosclearing */
-		BIOS_ZeroExtendedSize(false);
+XMS::XMS(Section* configuration) : Module_base(configuration), callbackhandler{}
+{
+	Section_prop* section = static_cast<Section_prop*>(configuration);
 
-		/* Remove Multiplex */
-		DOS_DelMultiplexHandler(multiplex_xms);
+	umb.is_available       = false;
+	a20.enable_global      = false;
+	a20.enable_local_count = 0;
 
-		/* Free used memory while skipping the 0 handle */
-		for (Bitu i = 1;i<XMS_HANDLES;i++) 
-			if(!xms_handles[i].free) XMS_FreeMemory(i);
+	if (!section->Get_bool("xms")) {
+		return;
 	}
 
-};
-static XMS* test;
+	// NTS: Disable XMS emulation if CPU type is less than a 286, because
+	// extended memory did not exist until the CPU had enough address lines
+	// to read past the 1MB mark.
+	//
+	// The other reason we do this is that there is plenty of software that
+	// assumes 286+ instructions if they detect XMS services, including but
+	// not limited to:
+	//
+	//      MSD.EXE Microsoft Diagnostics
+	//      Microsoft Windows 3.0
+	//
+	// Not emulating XMS for 8086/80186 emulation prevents the software
+	// from crashing.
+
+	if (CPU_ArchitectureType < CPU_ARCHTYPE_286) {
+		LOG_WARNING("XMS: CPU 80186 or lower lacks address lines needed for XMS, disabling");
+		return;
+	}
+
+	xms.is_available = true;
+	// TODO: read HMA configuration
+
+	BIOS_ZeroExtendedSize(true);
+	DOS_AddMultiplexHandler(xms_multiplex);
+
+	// Place hookable callback in writable memory area
+	xms.callback = RealMake(static_cast<uint16_t>(DOS_GetMemory(0x1) - 1),
+	                        0x10);
+	callbackhandler.Install(&XMS_Handler,
+	                        CB_HOOKABLE,
+	                        Real2Phys(xms.callback),
+	                        "XMS Handler");
+	// pseudocode for CB_HOOKABLE:
+	//	jump near skip
+	//	nop,nop,nop
+	//	label skip:
+	//	callback XMS_Handler
+	//	retf
+
+	for (uint16_t index = 0; index < NumXmsHandles; ++index) {
+		xms.handles[index] = XMS_Block();
+	}
+	xms.handles[0].is_free = false;
+
+	// Set up UMB chain
+	umb.is_available = section->Get_bool("umb");
+	const bool ems_available = GetEMSType(section) > 0;
+	DOS_BuildUMBChain(section->Get_bool("umb"), ems_available);
+
+	// TODO: If implementing CP/M compatibility, mirror the JMP
+	//       instruction in HMA
+}
+
+XMS::~XMS()
+{
+	// Remove upper memory information
+	dos_infoblock.SetStartOfUMBChain(0xffff);
```

Comment:
`0xffff` is used in a bunch of places - although it's UINT16_MAX, can is be turned into a const describing the last memory page? (if that's its logical meaning?)

Reply:
Unfortunately, I don't know. I didn't investigate the UMB functionality. Feel free to reopen if you have some idea.